src/main/java/software/amazon/awssdk/crt/mqtt/MqttConnection.java

```diff
@@ -127,130 +115,129 @@ public ConnectionState getState() {
+    public CompletableFuture<Void> connect(String clientId, boolean cleanSession) {
+        return connect(clientId, cleanSession, 0);
     }
 
-    public boolean connect(String clientId, boolean cleanSession, int keepAliveMs, MqttActionListener ack) {
+    public CompletableFuture<Void> connect(String clientId, boolean cleanSession, int keepAliveMs) {
         // Just clamp the keepAlive, no point in throwing
         short keepAlive = (short)Math.max(0, Math.min(keepAliveMs, Short.MAX_VALUE));
-        AsyncCallback connectAck = wrapAck(ack);
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        AsyncCallback connectAck = wrapVoidFuture(future);
         try {
             connectionState = ConnectionState.CONNECTING;
-            mqtt_connect(native_ptr(), clientId, cleanSession, keepAlive, connectAck);
+            mqttConnectionConnect(native_ptr(), clientId, cleanSession, keepAlive, connectAck);
         } catch (CrtRuntimeException ex) {
-            return false;
+            future.completeExceptionally(ex);
         }
-        return true;
+        return future;
     }
 
-    public void disconnect() {
-        disconnect(null);
-    }
-
-    public void disconnect(MqttActionListener ack) {
+    public CompletableFuture<Void> disconnect() {
+        CompletableFuture<Void> future = new CompletableFuture<>();
         if (native_ptr() == 0) {
-            return;
+            future.complete(null);
+            return future;
         }
-        AsyncCallback disconnectAck = wrapAck(ack);
+        AsyncCallback disconnectAck = wrapVoidFuture(future);
         connectionState = ConnectionState.DISCONNECTING;
-        mqtt_disconnect(native_ptr(), disconnectAck);
-    }
-
-    public short subscribe(String topic, QOS qos, Consumer<MqttMessage> handler) throws MqttException {
-        return subscribe(topic, qos, handler, null);
+        mqttConnectionDisconnect(native_ptr(), disconnectAck);
+        return future;
     }
 
-    public short subscribe(String topic, QOS qos, Consumer<MqttMessage> handler, MqttActionListener ack) throws MqttException {
+    public CompletableFuture<Integer> subscribe(String topic, QoS qos, Consumer<MqttMessage> handler) throws MqttException {
         if (native_ptr() == 0) {
             throw new MqttException("Invalid connection during subscribe");
         }
 
-        AsyncCallback subAck = wrapAck(ack);
+        CompletableFuture<Integer> future = new CompletableFuture<>();
+        AsyncCallback subAck = wrapPacketFuture(future);
         try {
-            return mqtt_subscribe(native_ptr(), topic, qos.getValue(), new MessageHandler(topic, handler), subAck);
+            int packetId = mqttConnectionSubscribe(native_ptr(), topic, qos.getValue(), new MessageHandler(topic, handler), subAck);
+            // When the future completes, complete the returned future with the packetId
+            return future.thenApply(unused -> packetId);
```

Comment:
In the future, there might be more we want to convey. Ex: QOS granted by the server (might be lower than user asked for). And I have no idea how multi-subscribe should be handled, where SOME of the topics might have succeeded.

Anyway, you could change this when we get those changes from the underlying libraries

Reply:
yeah, in that case, the return type will be more than packetId. Problem for future us :)